File: C#/myLeetCode/Binary_Tree/Traversal mode of Binary tree/iteration/144. Binary Tree Preorder Traversal.cs
// 144. Binary Tree Preorder Traversal
// https://leetcode.com/problems/binary-tree-preorder-traversal/

using System;
using System.Collections.Generic;

public class TreeNode {
    public int val;
    public TreeNode left;
    public TreeNode right;
    public TreeNode(int x) { val = x; }
}

public class Solution {
    public IList<int> PreorderTraversal(TreeNode root) {
        IList<int> result = new List<int>();
        if (root == null) return result;

        Stack<TreeNode> stack = new Stack<TreeNode>();
        stack.Push(root);

        while (stack.Count > 0) {
            TreeNode currentNode = stack.Pop();
            result.Add(currentNode.val);

            if (currentNode.right != null) {
                stack.Push(currentNode.right);
            }
            if (currentNode.left != null) {
                stack.Push(currentNode.left);
            }
        }

        return result;
    }
}

public class Program {
    public static void Main() {
        Solution solution = new Solution();

        TreeNode tree = new TreeNode(1) {
            right = new TreeNode(2) {
                left = new TreeNode(3)
            }
        };

        IList<int> preorderTraversal = solution.PreorderTraversal(tree);

        Console.WriteLine("Preorder Traversal:");
        foreach (int val in preorderTraversal) {
            Console.Write(val + " ");
        }
    }
}

File: C++/代码随想录/Binary_Tree/Traversal mode of Binary tree/recursion/144. Binary Tree Preorder Traversal.py
from typing import Optional
from typing import List

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        return [root.val] + self.preorderTraversal(root.left) + self.preorderTraversal(root.right) if root else []

sol = Solution()

# 测试用例
root = TreeNode(1, None, TreeNode(2, TreeNode(3)))
print(sol.preorderTraversal(root))  # 应输出 [1, 2, 3]
